cache/lru/lru.go

```diff
@@ -0,0 +1,173 @@
+// recently used element to evict was bigger than the size needed to evict.
+func (c *Cache) evict(sizeTotal uint64) error {
+	if sizeTotal > c.capacity {
+		return fmt.Errorf(
+			"can't evict %v elements in size, since capacity is %v",
+			sizeTotal, c.capacity)
+	}
+
+	evicted := uint64(0)
+	for evicted < sizeTotal {
+		// We still need to evict some more elements.
+		if c.ll.Len() == 0 {
+			// We should never reach here.
+			return fmt.Errorf(
+				"all elements got evicted, yet still need to evict %v,"+
+					"likelihood of error during size calculation",
+				sizeTotal-evicted)
+		}
+
+		// Find the least recently used item.
+		if elr := c.ll.Back(); elr != nil {
+			// Determine lru item's size.
+			ce := elr.Value.(*entry)
+			es, err := ce.value.Size()
+			if err != nil {
+				return fmt.Errorf(
+					"couldn't determine size of existing cache value %v", err)
+			}
+
+			// Account for that element's removal in evicted and cache size.
+			evicted += es
+			c.size -= es
+
+			// Remove the element from the cache.
+			c.ll.Remove(elr)
+			delete(c.cache, ce.key)
+		}
+	}
+
+	return nil
+}
+
+// Put inserts a given (key,value) pair into the cache, if the key already
+// exists, it will replace value and update it to be most recent item in cache.
+func (c *Cache) Put(key interface{}, value cache.Value) error {
+	vs, err := value.Size()
+	if err != nil {
+		return fmt.Errorf("couldn't determine size of cache value: %v", err)
+	}
+	if vs > c.capacity {
+		return fmt.Errorf(
+			"can't insert entry of size %v into cache with capacity %v",
+			vs, c.capacity)
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	// If the filter already exists, remove it and decrease cache's size.
+	el, ok := c.cache[key]
+	if ok {
+		es, err := el.Value.(*entry).value.Size()
+		if err != nil {
+			return fmt.Errorf(
+				"couldn't determine size of existing cache value %v", err)
+		}
+		c.ll.Remove(el)
+		c.size -= es
+	}
+
+	// Then we need to make sure we have enough space for the filter, evict
+	// filters if we need more space.
+	if c.size+vs > c.capacity {
+		err := c.evict(c.size + vs - c.capacity)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We have made enough space in the cache, so just insert it.
+	el = c.ll.PushFront(&entry{key, value})
+	c.cache[key] = el
+	c.size += vs
+
+	return nil
+}
+
+// Get will return value for a given key, making the element the most recently
+// accessed item in the process. Will return nil if the key isn't found.
+func (c *Cache) Get(key interface{}) (cache.Value, error) {
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+
+	el, ok := c.cache[key]
+	// Element not found in the cache.
+	if !ok {
+		return nil, ErrElementNotFound
+	}
+
+	// When the cache needs to evict a filter to make space for another one, it
+	// starts eviction from the back, so by moving this filter to the front,
+	// it's eviction is delayed because it's recently accessed.
+	c.ll.MoveToFront(el)
```

Comment:
We must actually acquire a write lock before calling this (I encountered this running race condition tests).

Reply:
great catch! also added a test case to cover it.